### ISSUE TITLE
Feature: constraining scale domains on locus datatype

### ISSUE THREAD

**Issue by alexpreynolds** [tuner/genome-spy#27]
Hello, I have been playing with a fork of your ideogram track code from here: https://observablehq.com/@tuner/annotation-tracks

My goal is to be able to generate a specification object that constrains the view to a genomic start and end positions, e.g. such as `chr1:100000` to `chr3:123456`, using the chromosomal ordering and sizes provided in an assembly's cytoBand track file, for instance.

The reason for this is that I have a separate Observable notebook with code that I can readily use to lookup intervals, such as those coming from gene or other custom annotation tables. It would be useful to provide an interval to the specification to "zoom in" to a given region of interest.

I have tried a couple things and am unsure if this is an issue, or if I am specifying things incorrectly. Probably the latter.

The starting view for the `hg38` assembly looks like this:

<img width="933" alt="Screen Shot 2021-07-17 at 3 35 16 PM" src="https://user-images.githubusercontent.com/33584/126050736-770c7312-63d2-4db1-a1ef-2cdf5b4ee6be.png">

The axis is not shown, but the default ideogram renders all chromosomes in natural sort order (chr1, chr2, etc.).

The first attempt was to simply view all of `chr1`, by applying a `filter` transform on the `chrom` field:

```
  transform: [
    // Remove unlocalized/unplaced scaffolds etc.
    { type: "filter", expr: "!test(/_/, datum.chrom)" },
    { type: "filter", expr: "datum.chrom === 'chr1'" }
  ],
```

The code for this version is at: https://observablehq.com/d/e085ea04064331ed

While this only renders bands for chromosome 1, the larger view reserves horizontal space for all chromosomes (I removed the `axis` property to show the chromosome boundaries):

<img width="933" alt="Screen Shot 2021-07-17 at 3 36 31 PM" src="https://user-images.githubusercontent.com/33584/126050773-51aa7c2f-7d57-42c9-9027-a56393e359ec.png">

The second attempt was to specify a `scale.domain` property to clip the loci by the `chrom` field to `chr1`, similar to how I would clip a bar chart with a nominal axis in the Vega test editor:

```
  encoding: {
    x: {
      chrom: "chrom",
      pos: "chromStart",
      type: "locus",
      scale: {
        domain: [{ chrom: "chr1" }]
      }
    },
    ...
  },
```

The code for this version is at: https://observablehq.com/d/999e38b6bb6d08f8

This did not clip the dataset:

<img width="1182" alt="Screen Shot 2021-07-17 at 3 45 05 PM" src="https://user-images.githubusercontent.com/33584/126050887-9f321d97-9ff7-46ce-b4a2-c5b458c79fd2.png">

Do I need to constrain the genomic range via values from the `linearizeGenomicCoordinate` function, or use a different approach? https://github.com/tuner/genome-spy/blob/c126dcdeb45294a1747dcc85d03ec7274f0be68c/docs/grammar/scale.md#locus-scale

Thanks for making this project and for any guidance you can provide.

**Comment by tuner** [comment]
Hi Alex,

I haven't had a use case for constraining the genomic domain, so it may currently be difficult or impossible. In principle, you should be able to constrain the domain by using linearized coordinates (https://github.com/tuner/genome-spy/blob/master/packages/genome-spy/src/view/scaleResolution.js#L146). However, it seems to do nothing.

Anyway, this is something that should be supported because not everyone is interested in whole genomes. Also, having means for specifying an initial zoomed domain could be useful. What you tried (`domain: [{ chrom: "chr1" }]`) could be a convenient way to specify genomic domains, although there are some details that need to be figured out first. There should also be an option to extract the domain from the data  (actually, this the default behavior on non-genomic scales), e.g., if all the features are at the end of chr1, only display that region. Also, some clever padding should be added to provide context (somewhat analogous to rounding to the domain to nice numbers).

I'm on vacation for two more weeks and trying to avoid writing any code. I "accidentally" opened github! I could come up with some quick fix for your particular use case when I'm back at work.

Let me know if you have more needs or ideas related to domain handling! It's best to discuss this more broadly. Ideally, we will have a unified, generic solution instead of multiple ad hoc options.

**Comment by alexpreynolds** [comment]
Thanks for getting back and sorry to interrupt your holidays!

Zooming into a specified range would be useful for looking at genes of interest. Reporting the current genomic range within-view would also be helpful for some visualizations I have in mind.

As an example, if I can get back the current bounds within-view, I could build a query to an existing higlass server backend, which returns a generic matrix of floats. I would use those floats for building a stacked bar chart (rebuilding an epilogos figure, basically).

To that end, I've been looking at how mouse events work in GenomeSpy, which could be used to report back details about the current state of the view.

Perhaps we can touch base when you return and have some free time. We can discuss more generic ways to get and change genomic bounds or other view state.

Until then, please enjoy your time off and have a great summer!

**Comment by mattjvincent** [comment]
I am interested in this as well.  Being able to specify a region of the genome would be very useful indeed.

The following is **NOT** the correct answer, but it will draw just chromosome 1.

```
viewof genomeSpyIdeogram = embed({
  genome: {
        "name": "Test Genome",
        "contigs": [
          {"name": "chr1", "size": 247249719},
        ]    
  },
  vconcat: [makeIdeogramTrack("hg38")]
})
```

[Observable notebook](https://observablehq.com/d/baaa9e23c8081260)

**Comment by tuner** [comment]
I opened an issue for the observable scales: https://github.com/tuner/genome-spy/issues/29

**Comment by tuner** [comment]
Actually, specifying the domain using linearized cordinates seems to work:

```json
{
  "genome": { "name": "hg38" },
  "data": { "values": [{ "chr": 4, "pos": 50000000 }] },
  "mark": "point",
  "encoding": {
    "x": {
      "chrom": "chr",
      "pos": "pos",
      "type": "locus",
      "scale": { "domain": [491149951, 1061198324] }
    }
  }
}
````

## Specifying domains (and intervals in general) using genomic coordinates

Because this is quite inconvenient, the domain could be specified using genomic coordinates. However, there are some tricky details that need to be figured out.

### Exact coordinates, chromosomes 3, 4, and 5:
```json
[{ "chrom": "chr3", "pos": 0 }, {"chrom": "chr5", "pos": 181538259 }]
```

This still needs the size of the last chromosome, which may be inconvenient. 

### Exact coordinates, chromosomes 3, 4, and 5. Employing half-open intervals:
```json
[{ "chrom": "chr3", "pos": 0 }, {"chrom": "chr6", "pos": 0 }]
```

The first base of chr6 is not included. But oops, how to include the last chromosome of the assembly with this method? The next method could be a solution.

### Just chromosomes 3, 4, and 5, no positions:
```json
[{ "chrom": "chr3" }, {"chrom": "chr5" }]
```

The left endpoint is obvious, it's `chr3:0`. But how should the right endpoint be treated, is it `chr5:0` or `chr5:181538259`? I think that it must include the whole last chromosome. Otherwise this method could not be used to specify a domain that spans the whole genome. 

So, the `pos` property would be optional. The position on the left endpoint would default to zero and the right endpoint would default to the length of the chromosome.

### A single chromosome:

```json
[{ "chrom": "chr3" }, {"chrom": "chr3" }]
```

## Other use cases

This could also be applied to, for instance, the initial zoom region.

### Encoding data

It currently tricky to specify encoding for elements that span whole chromosomes. An example of such case is an alternating background color:

![image](https://user-images.githubusercontent.com/399972/127980215-74182f84-3746-4d14-83ac-ec149151329f.png)

Thus if the same default-position logic is used in encodings, the above background could be specified this way:

```json
{
  "genome": { "name": "hg38" },
  "data": { "values": ["chr1", "chr3", "chr5", ...] },
  "mark": "rect",
  "encoding": {
    "x": { "chrom": "data", "type": "locus" },
    "x2": { "chrom": "data" }
  }
}

```

**Comment by tuner** [comment]
# Initial zoom

In GenomeSpy the scales can be made zoomable using the `zoom` property. Zooming and panning are constrained to the specified (or exctracted-from-data) domain:

```json
"scale": {
  "domain": [0, 100],
  "zoom": true
}
```

Generally, I'd like to be conservative and maintain compatibility with Vega-Lite, but they haven't solved this either. In Vega-Lite the zoom interaction requires selections and doesn't support constraining: https://vega.github.io/vega-lite/examples/selection_translate_scatterplot_drag.html, https://github.com/vega/vega-lite/issues/4886. Also, in Vega-Lite `domain` specifies the initial domain whereas in GenomeSpy it specifies both the initial domain and the extent of zoom/pan.

Probably the most straightforward way to allow specifying an initial zoom that is different from the constraint would be to allow objects as values in the `zoom` property:

```json
"scale": {
  "domain": [0, 100],
  "zoom": {
    "initialDomain": [40, 60]
  }
}
```

`initialDomain` should also accept chrom/pos objects.

An alternative method would be to use the (still non-existing) API to lookup the scale resolution and call the `zoomTo` method  once the visualization has been initialized.

**Comment by alexpreynolds** [comment]
I tested out a domain with linearized coordinates in the current version of GenomeSpy and the example specification draws a point between chr3 and chr5, as you mention:

<img width="962" alt="Screen Shot 2021-08-04 at 4 51 17 PM" src="https://user-images.githubusercontent.com/33584/128269965-7503de3f-7a68-47eb-8002-b4501a210cbe.png">

I tried it out on an Observable notebook using the ideogram track, but this did not appear to use the custom domain bounds: https://observablehq.com/d/5fa3e690273a6d14

In the example above, I also tried `scale: { domain: [0, 248956422] }` to get `chr1` from `hg38` assembly. 

I also tried something similar in the paired Manhattan composite track (https://observablehq.com/d/f465306968baebef) but that did not change the layout.

One option could be a helper function that draws an hg38-specific chromosome sizes file from the web (or reuses what is obtained for the ideogram) and parses that into cumulative sizes for the set of displayed chromosomes. Those can then be used to convert genomic coordinates into "absolute" or linear coordinates to place into the specification.

**Comment by tuner** [comment]
Dang, there was a bug after all: https://github.com/tuner/genome-spy/commit/e19c8363909bddf3794b87236db95acd39af0b91

I released a new version and your notebook seems to work as expected.

The linked draft PR will allow you to use chromosomes to specify the domain. I'll try to implement the zoom extent option before merging it to master.

**Comment by tuner** [comment]
Here's the documentation: https://genomespy.app/docs/grammar/scale.html

I also published a new version (0.1.8) and updated it to Observable.

**Comment by alexpreynolds** [comment]
Thanks very much for the added features — the zoom extent works great. I was able to integrate a rough gene annotation lookup fairly quickly: https://observablehq.com/d/f465306968baebef
